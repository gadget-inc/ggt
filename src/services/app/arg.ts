import { ArgError } from "../error/error.js";
import { sprint } from "../output/sprint.js";

/**
 * Parses the value of an application argument (-a/--app) and returns
 * the application's slug.
 *
 * @param value - The value of the argument.
 * @param name - The name of the argument. e.g. "-a" or "--app".
 * @returns The application's slug.
 * @throws {ArgError} If the value is not a valid slug or URL.
 */
export const AppArg = (value: string, name: string): string => {
  const slug = /^(https:\/\/)?(?<slug>[\w-]+?)(--development)?(\..*)?$/.exec(value)?.groups?.["slug"];
  if (slug) {
    return slug;
  }

  throw new ArgError(
    sprint`
      ${name} must be the application's {bold slug} or {bold URL}

      {bold EXAMPLES:}
        ${name} my-app
        ${name} my-app.gadget.app
        ${name} https://my-app.gadget.app
        ${name} https://my-app.gadget.app/edit
        ${name} https://my-app--development.gadget.app/edit
    `,
  );
};
