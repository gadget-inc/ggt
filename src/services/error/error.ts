import arg from "arg";
import cleanStack from "clean-stack";
import type { GraphQLError } from "graphql";
import assert from "node:assert";
import { randomUUID } from "node:crypto";
import { dedent } from "ts-dedent";
import type { JsonObject } from "type-fest";
import type { CloseEvent, ErrorEvent } from "ws";
import type { Query } from "../app/edit-graphql.js";
import { env } from "../config/env.js";
import { sprintln } from "../output/sprint.js";
import { compact, uniq } from "../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString } from "../util/is.js";
import { serializeError } from "../util/object.js";

export enum IsBug {
  YES = "yes",
  NO = "no",
  MAYBE = "maybe",
}

/**
 * Base class for all errors.
 */
export abstract class CLIError extends Error {
  /**
   * A GGT_CLI_SOMETHING human/machine readable unique identifier for this error.
   */
  code: string;

  /**
   * The Sentry event ID for this error.
   */
  sentryEventId = env.testLike ? "00000000-0000-0000-0000-000000000000" : randomUUID();

  /**
   * The underlying *thing* that caused this error.
   */
  cause?: unknown;

  /**
   * Assume the stack trace exists.
   */
  override stack!: string;

  /**
   * Indicates whether this error is considered a bug or not.
   */
  abstract isBug: IsBug;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    Error.captureStackTrace(this, this.constructor);
  }

  /**
   * Constructs a CLIError from a cause.
   */
  static from(cause: unknown): CLIError {
    if (cause instanceof CLIError) {
      return cause;
    }
    if (cause instanceof arg.ArgError) {
      return new ArgError(cause.message);
    }
    return new UnexpectedError(cause);
  }

  /**
   * Turns this error into a user-friendly message that explains what
   * went wrong and how to fix it. A good write up of what an error
   * should look like can be found here:
   * {@link https://clig.dev/#errors}
   */
  render(): string {
    return compact([this.header(), this.body(), this.footer()]).join("\n\n");
  }

  protected header(): string {
    return `${this.code}: ${this.message}`;
  }

  protected footer(): string {
    if (this.isBug === IsBug.NO) {
      return "";
    }

    return dedent`
      ${this.isBug === IsBug.YES ? "This is a bug" : "If you think this is a bug"}, please submit an issue using the link below.

      https://github.com/gadget-inc/ggt/issues/new?template=bug_report.yml&error-id=${this.sentryEventId}
    `;
  }

  protected abstract body(): string;
}

/**
 * Our "catch all" error.
 *
 * If this error is thrown, we almost certainly have a bug, and should
 * either fix it or add a more specific error so that we can provide
 * more useful information.
 */
export class UnexpectedError extends CLIError {
  isBug = IsBug.YES;

  constructor(override cause: unknown) {
    super("GGT_CLI_UNEXPECTED_ERROR", "An unexpected error occurred");
  }

  protected body(): string {
    if (isError(this.cause)) {
      return cleanStack(this.cause.stack ?? this.stack);
    }
    return this.stack;
  }
}

export class EditGraphQLError extends CLIError {
  isBug = IsBug.MAYBE;

  override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly query: Query<JsonObject>,
    cause: unknown,
  ) {
    super("GGT_CLI_CLIENT_ERROR", "An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override body(): string {
    if (isGraphQLErrors(this.cause)) {
      const errors = uniq(this.cause.map((x) => x.message));
      if (errors.length > 1) {
        let n = 1;
        return sprintln("Gadget responded with multiple errors:").concat(`\n  ${n++}. ${errors.join(`\n  ${n++}. `)}`);
      } else {
        return dedent`
          Gadget responded with the following error:

            ${errors[0]}
        `;
      }
    }

    if (isCloseEvent(this.cause)) {
      return "The connection to Gadget closed unexpectedly.";
    }

    if (isErrorEvent(this.cause) || isError(this.cause)) {
      return this.cause.message;
    }

    return this.cause;
  }
}

export class YarnNotFoundError extends CLIError {
  isBug = IsBug.NO;

  constructor() {
    super("GGT_CLI_YARN_NOT_FOUND", "Yarn not found");
  }

  protected body(): string {
    return dedent`
      Yarn must be installed to sync your application. You can install it by running:

        $ npm install --global yarn

      For more information, see: https://classic.yarnpkg.com/en/docs/install
    `;
  }
}

export class ArgError extends CLIError {
  isBug = IsBug.NO;

  constructor(message: string) {
    super("GGT_CLI_ARG_ERROR", message);
  }

  // eslint-disable-next-line lodash/prefer-constant
  protected override header(): string {
    return "";
  }

  protected override body(): string {
    return this.message;
  }
}

export class InvalidSyncFileError extends CLIError {
  isBug = IsBug.MAYBE;

  constructor(
    readonly dir: string,
    readonly app: string | undefined,
  ) {
    super("GGT_CLI_INVALID_SYNC_FILE", "The .gadget/sync.json file was invalid or not found");
    this.app ??= "<name of app>";
  }

  protected body(): string {
    return dedent`
      We failed to read the Gadget metadata file in this directory:

        ${this.dir}

      If you're running \`ggt sync\` for the first time, we recommend using an empty directory such as:

        ~/gadget/${this.app}

      Otherwise, if you're sure you want to sync the contents of that directory to Gadget, run \`ggt sync\` again with the \`--force\` flag:

        $ ggt sync --app ${this.app} ${this.dir} --force

      You will be prompted to either merge your local files with your remote ones or reset your local files to your remote ones.
    `;
  }
}
