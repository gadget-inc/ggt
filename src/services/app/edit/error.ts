import type { GraphQLError } from "graphql";
import assert from "node:assert";
import pluralize from "pluralize";
import type { CloseEvent, ErrorEvent } from "ws";
import { sprint } from "../../output/print.js";
import { CLIError, IsBug } from "../../output/report.js";
import { uniq } from "../../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString } from "../../util/is.js";
import { serializeError } from "../../util/object.js";
import type { GraphQLMutation, GraphQLQuery, GraphQLSubscription } from "./operation.js";

export class EditError extends CLIError {
  isBug = IsBug.MAYBE;

  override cause: string | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly request: GraphQLQuery | GraphQLMutation | GraphQLSubscription,
    cause: unknown,
  ) {
    super("An error occurred while communicating with Gadget");

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override render(): string {
    let body = "";

    switch (true) {
      case isGraphQLErrors(this.cause): {
        const errors = uniq(this.cause.map((x) => x.message));
        body = sprint`
          Gadget responded with the following ${pluralize("error", errors.length, false)}:

            • ${errors.join("\n            • ")}
        `;
        break;
      }
      case isCloseEvent(this.cause):
        body = "The connection to Gadget closed unexpectedly.";
        break;
      case isErrorEvent(this.cause) || isError(this.cause):
        body = this.cause.message;
        break;
      default:
        body = this.cause;
        break;
    }

    return this.message + "\n\n" + body;
  }
}
