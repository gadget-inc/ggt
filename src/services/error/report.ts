import * as Sentry from "@sentry/node";
import ms from "ms";
import os from "node:os";
import { serializeError } from "serialize-error";
import type { App } from "../app/app.js";
import { config } from "../config/config.js";
import { env } from "../config/env.js";
import { createLogger } from "../output/log/logger.js";
import type { User } from "../user/user.js";
import { parseBoolean } from "../util/boolean.js";
import { CLIError, IsBug } from "./error.js";

const log = createLogger({ name: "report" });

let user: User | undefined;
export const setUser = (newUser: typeof user): void => {
  log.trace("set user", { user: newUser });
  user = newUser;
  // eslint-disable-next-line unicorn/no-null
  Sentry.setUser(newUser ?? null);
};

let app: App | undefined;
export const setApp = (newApp: typeof app): void => {
  log.trace("set app", { app: newApp });
  app = newApp;
};

export const reportErrorAndExit = async (cause: unknown): Promise<never> => {
  log.error("reporting error and exiting", { error: cause });

  try {
    const error = CLIError.from(cause);
    log.println(error.render());

    if (error.isBug !== IsBug.NO) {
      Sentry.getCurrentHub().captureException(error, {
        event_id: error.sentryEventId,
        captureContext: {
          user: user ? { id: String(user.id), email: user.email, username: user.name ?? undefined } : undefined,
          tags: {
            applicationId: app ? app.id : undefined,
            arch: config.arch,
            bug: error.isBug,
            code: error.code,
            environment: env.value,
            platform: config.platform,
            shell: config.shell,
            version: config.version,
          },
          contexts: {
            cause: error.cause ? serializeError(error.cause) : undefined,
            app: {
              command: `ggt ${process.argv.slice(2).join(" ")}`,
              argv: process.argv,
            },
            device: {
              name: os.hostname(),
              family: os.type(),
              arch: os.arch(),
            },
            runtime: {
              name: process.release.name,
              version: process.version,
            },
          },
        },
      });

      await Sentry.flush(ms("2s"));
    }
  } finally {
    process.exit(1);
  }
};

export const installErrorHandlers = (): void => {
  log.debug("installing error handlers");

  Sentry.init({
    dsn: "https://0c26e0d8afd94e77a88ee1c3aa9e7065@o250689.ingest.sentry.io/6703266",
    release: config.version,
    enabled: env.productionLike && parseBoolean(process.env["GGT_SENTRY_ENABLED"] ?? "true"),
  });

  process.once("uncaughtException", (error) => {
    void reportErrorAndExit(error);
  });

  process.once("unhandledRejection", (error) => {
    void reportErrorAndExit(error);
  });
};
