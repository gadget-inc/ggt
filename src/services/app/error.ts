import type { GraphQLError } from "graphql";
import assert from "node:assert";
import pluralize from "pluralize";
import type { CloseEvent, ErrorEvent } from "ws";
import { GGTError, IsBug } from "../output/report.js";
import { sprint } from "../output/sprint.js";
import { uniq } from "../util/collection.js";
import { isCloseEvent, isError, isErrorEvent, isGraphQLErrors, isString, isStringArray } from "../util/is.js";
import { serializeError } from "../util/object.js";
import type { GraphQLMutation, GraphQLQuery, GraphQLSubscription } from "./edit/operation.js";

export class ClientError extends GGTError {
  public isBug: IsBug;

  override cause: string | string[] | Error | readonly GraphQLError[] | CloseEvent | ErrorEvent;

  constructor(
    readonly request: GraphQLQuery | GraphQLMutation | GraphQLSubscription,
    cause: unknown,
    isBug?: IsBug,
  ) {
    super("An error occurred while communicating with Gadget");

    if (isBug) {
      this.isBug = isBug;
    } else {
      this.isBug = IsBug.MAYBE;
    }

    // ErrorEvent and CloseEvent aren't serializable, so we reconstruct
    // them into an object. We discard the `target` property because
    // it's large and not that useful
    if (isErrorEvent(cause)) {
      this.cause = {
        type: cause.type,
        message: cause.message,
        error: serializeError(cause.error),
      } as ErrorEvent;
    } else if (isCloseEvent(cause)) {
      this.cause = {
        type: cause.type,
        code: cause.code,
        reason: cause.reason,
        wasClean: cause.wasClean,
      } as CloseEvent;
    } else {
      assert(
        isString(cause) || isStringArray(cause) || isError(cause) || isGraphQLErrors(cause),
        "cause must be a string, Error, GraphQLError[], CloseEvent, or ErrorEvent",
      );
      this.cause = cause;
    }
  }

  override render(): string {
    let body: string;

    switch (true) {
      case isGraphQLErrors(this.cause): {
        const errors = uniq(this.cause.map((x) => x.message));
        body = sprint`
          Gadget responded with the following ${pluralize("error", errors.length, false)}:

            • ${errors.join("\n            • ")}
        `;
        break;
      }
      case isCloseEvent(this.cause):
        body = "The connection to Gadget closed unexpectedly.";
        break;
      case isErrorEvent(this.cause) || isError(this.cause):
        body = this.cause.message;
        break;
      case isStringArray(this.cause):
        if (this.cause.length === 1) {
          body = String(this.cause[0]);
        } else {
          body = this.cause.join(", ");
        }
        break;
      default:
        body = this.cause;
        break;
    }

    return this.message + "\n\n" + body;
  }
}

export class AuthenticationError extends ClientError {
  constructor(request: GraphQLQuery | GraphQLMutation | GraphQLSubscription) {
    super(request, "Request authentication failed due to the session expiring while running the command. Please sign-in again.", IsBug.NO);
  }
}
