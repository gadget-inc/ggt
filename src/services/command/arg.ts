import arg from "arg";
import type { Simplify } from "type-fest";
import { GGTError, IsBug, UnexpectedError } from "../output/report.js";
import { isNil } from "../util/is.js";

export type ArgsDefinition = Record<string, ArgDefinition>;

type ArgDefinition<Handler extends arg.Handler = arg.Handler> =
  | Handler
  | {
      type: Handler;
      alias?: string | string[];
      default?: ReturnType<Handler>;
    };

export type ParseArgsOptions = {
  /**
   * A list of arguments to parse.
   */
  argv?: string[];

  /**
   * When permissive set to `true`, arg will push any unknown arguments
   * onto the "extra" argument array (`ctx.args._`) instead of throwing
   * an error about an unknown flag.
   *
   * @default false
   */
  permissive?: boolean;

  /**
   * When stopAtPositional is set to true, context will stop parsing at
   * the first positional argument.
   *
   * @default false
   */
  stopAtPositional?: boolean;
};

export const parseArgs = <Args extends ArgsDefinition>(args: Args, options?: arg.Options): ArgsDefinitionResult<Args> => {
  const spec: arg.Spec = {};
  const defaultValues: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(args)) {
    if (!("type" in value)) {
      spec[key] = value;
      continue;
    }

    spec[key] = value.type;
    defaultValues[key] = value.default;

    if (value.alias) {
      for (const alias of Array.isArray(value.alias) ? value.alias : [value.alias]) {
        spec[alias] = key;
      }
    }
  }

  try {
    const parsed = arg(spec, options);
    for (const [key, value] of Object.entries(defaultValues)) {
      if (isNil(parsed[key])) {
        parsed[key] = value as never;
      }
    }
    return parsed as ArgsDefinitionResult<Args>;
  } catch (error: unknown) {
    if (error instanceof arg.ArgError) {
      // convert arg.ArgError to GGTError
      // eslint-disable-next-line no-ex-assign
      error = new ArgError(error.message);
    }
    if (error instanceof GGTError) {
      throw error;
    }
    throw new UnexpectedError(error);
  }
};

export class ArgError extends GGTError {
  isBug = IsBug.NO;

  protected override render(): string {
    return this.message;
  }
}

/**
 * Turns this:
 * ```ts
 * type Args = {
 *   "--string": { type: String; alias: "s" };
 *   "--number": { type: Number; default: 42 };
 * };
 * ```
 *
 * Into this:
 * ```ts
 * type Result = {
 *  "--string": string | undefined;
 *  "--number": number;
 * };
 * ```
 */
export type ArgsDefinitionResult<Args extends ArgsDefinition, Keys extends keyof Args = keyof Args> = Simplify<{
  [Key in Keys]: Args[Key] extends ArgDefinition<infer Handler>
    ? Args[Key] extends { default: unknown }
      ? NonNullable<ReturnType<Handler>>
      : ReturnType<Handler> | undefined
    : never;
}> & { _: string[] };
