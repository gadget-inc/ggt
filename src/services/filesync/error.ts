import { ClientError } from "../app/error.js";
import type { Context } from "../command/context.js";
import { sprintProblems, type Problems } from "../output/problems.js";
import { GGTError, IsBug } from "../output/report.js";
import { sprint, sprintln } from "../output/sprint.js";
import { isGraphQLErrors, isGraphQLResult, isObject, isString } from "../util/is.js";
import type { Directory } from "./directory.js";
import type { SyncJsonArgs } from "./sync-json.js";

export class YarnNotFoundError extends GGTError {
  isBug = IsBug.NO;

  constructor() {
    super("Yarn not found");
  }

  protected render(): string {
    return sprint`
      Yarn must be installed to sync your application. You can install it by running:

        $ npm install --global yarn

      For more information, see: https://classic.yarnpkg.com/en/docs/install
    `;
  }
}

export class UnknownDirectoryError extends GGTError {
  isBug = IsBug.NO;

  constructor(
    readonly ctx: Context<SyncJsonArgs>,
    readonly opts: { directory: Directory },
  ) {
    super('The ".gadget/sync.json" file was invalid or not found');
  }

  protected render(): string {
    const cmd = this.ctx.command;
    const dir = this.opts.directory.path;

    switch (cmd) {
      case "open":
      case "status":
        return sprint`
          A ".gadget/sync.json" file is missing in this directory:

            ${dir}

          In order to use "ggt ${cmd}", you must run it within a directory
          that has already been initialized with "ggt dev".
        `;
      case "dev":
        return sprint`
          A ".gadget/sync.json" file is missing in this directory:

            ${dir}

          If you're running "ggt dev" for the first time, we recommend
          using a gadget specific directory like this:

            ggt dev ~/gadget/${this.ctx.args["--app"] ?? "<name>"}

          To use a non-empty directory without a ".gadget/sync.json" file,
          run "ggt dev" again with the "--allow-unknown-directory" flag:

            ggt dev ${dir} --allow-unknown-directory
        `;
      default:
        return sprint`
          A ".gadget/sync.json" file is missing in this directory:

            ${dir}

          If you're certain you want to use this directory, you can run
          "ggt ${cmd}" again with the "--allow-unknown-directory" flag:

            ggt ${cmd} --allow-unknown-directory
        `;
    }
  }
}

export class TooManyMergeAttemptsError extends GGTError {
  isBug = IsBug.MAYBE;

  constructor(readonly attempts: number) {
    super(`Failed to synchronize files after ${attempts} attempts.`);
  }

  protected render(): string {
    return sprint`
      We merged your local files with your environment's files ${this.attempts} times,
      but your local and environment's files still don't match.

      Make sure no one else is editing files on your environment, and try again.
    `;
  }
}

export class DeployDisallowedError extends GGTError {
  isBug = IsBug.MAYBE;

  constructor(readonly fatalErrors: Problems) {
    super("This application is not allowed to be deployed due to fatal errors.");
  }

  protected render(): string {
    let output = sprintln`{red Gadget has detected the following fatal errors with your files:}`;
    output += sprintProblems({ ensureEmptyLineAbove: true, problems: this.fatalErrors, showFileTypes: false });
    output += sprintln({ ensureEmptyLineAbove: true })`{red Please fix these errors and try again.}`;
    return output;
  }
}

export const isFilesVersionMismatchError = (error: unknown): boolean => {
  if (error instanceof ClientError) {
    error = error.cause;
  }
  if (isGraphQLResult(error)) {
    error = error.errors;
  }
  if (isGraphQLErrors(error)) {
    error = error[0];
  }
  return isObject(error) && "message" in error && isString(error.message) && error.message.includes("Files version mismatch");
};

export const swallowFilesVersionMismatch = (ctx: Context, error: unknown): void => {
  if (isFilesVersionMismatchError(error)) {
    ctx.log.debug("swallowing files version mismatch", { error });
    return;
  }
  throw error;
};
