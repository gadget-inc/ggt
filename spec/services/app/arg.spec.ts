import { describe, expect, it } from "vitest";
import { AppArg } from "../../../src/services/app/arg.js";
import { ArgError } from "../../../src/services/command/arg.js";
import { expectError } from "../../__support__/error.js";

describe("AppArg", () => {
  it.each([
    "my-app",
    "my-app.gadget.app",
    "my-app--development.gadget.app",
    "https://my-app.gadget.app",
    "https://my-app--development.gadget.app",
    "https://my-app.gadget.app/edit",
    "https://my-app.gadget.app/edit/files/routes/GET.js",
  ])("accepts %s", (value) => {
    expect(AppArg(value, "--app")).toEqual("my-app");
  });

  it.each(["~"])("rejects %s", async (value) => {
    const error = await expectError(() => AppArg(value, "--app"));
    expect(error).toBeInstanceOf(ArgError);
    expect(error.message).toMatchSnapshot();
  });
});
