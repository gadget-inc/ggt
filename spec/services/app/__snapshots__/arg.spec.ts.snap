// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppArg > rejects ~ 1`] = `
"app must be the application's slug or URL

EXAMPLES:
  app my-app
  app my-app.gadget.app
  app https://my-app.gadget.app
  app https://my-app.gadget.app/edit
  app https://my-app--development.gadget.app/edit"
`;
